src/components/fields/SchemaField.js

```diff
@@ -128,38 +129,50 @@ DefaultTemplate.defaultProps = {
   displayLabel: true,
 };
 
-function SchemaField(props) {
-  const {uiSchema, errorSchema, idSchema, name, required, registry} = props;
-  const {definitions, fields, formContext, FieldTemplate = DefaultTemplate} = registry;
-  const schema = retrieveSchema(props.schema, definitions);
-  const FieldComponent = getFieldComponent(schema, uiSchema, fields);
-  const {DescriptionField} = fields;
-  const disabled = Boolean(props.disabled || uiSchema["ui:disabled"]);
-  const readonly = Boolean(props.readonly || uiSchema["ui:readonly"]);
-  const autofocus = Boolean(props.autofocus || uiSchema["ui:autofocus"]);
-
-  if (Object.keys(schema).length === 0) {
-    // See #312: Ensure compatibility with old versions of React.
-    return <div/>;
+class SchemaField extends Component {
+  shouldComponentUpdate(nextProps, nextState) {
+    let _props = {...this.props};
+    let _nextProps = {...nextProps};
+    // if schema and formData are the equal
+    // idSchema and errorSchema will be equal
+    delete _props.idSchema;
+    delete _nextProps.idSchema;
+    return !deepEquals(_props, _nextProps);
```

Comment:
I'm confused with this code:

1. the implementation doesn't reflect what the comment says; we're not dropping `errorSchema`;
1. the comment suggests we should only compare changes for `schema` and `formData`, which is not what we're doing;
1. I suspect we should take care of `uiSchema` changes as well.

I'm trying to summarize the points above in a sample implementation below:

```jsx
shouldComponentUpdate(nextProps) {
  return !deepEquals(this.props.schema, nextProps.schema) ||
         !deepEquals(this.props.uiSchema, nextProps.uiSchema) ||
         !deepEquals(this.props.formData, nextProps.formData);
}
```

We may obviously want to ensure we're not missing other necessary checks, but I'm finding this way much more explicit.